File: scripts/check-workspace-assertions.py
#!/usr/bin/env python3
"""Sharp workspace architecture assertions for velvet-ballastics."""

from __future__ import annotations

import sys
import tomllib
from pathlib import Path


ACTIVE_MEMBERS = frozenset(
    {
        "crates/vb_boundary_inventory",
        "crates/vb_core",
        "crates/vb_yaml",
        "crates/vb_validate",
        "crates/vb_expr",
        "crates/vb_compile",
        "crates/vb_storage",
        "crates/vb_runtime",
        "crates/vb_doc",
        "crates/vb_ipc",
        "crates/workspace_tests",
    }
)

DEFERRED_MEMBERS = frozenset(
    {
        "crates/vb_codegen",
        "crates/vb_ui_makepad",
        "crates/vb_ui_snapshot",
        "crates/vb_proof_kernels",
        "crates/vb_cli",
        "crates/vb_verification",
        "crates/vb_benchmark",
        "fuzz",
    }
)

EXPECTED_EXCLUDES = frozenset({"target/miri-tmp", "crates/vb_ui", "fuzz"})
BOUNDARY_CRATES = frozenset({"vb_core", "vb_runtime", "vb_storage", "vb_ipc"})
FORBIDDEN_UI_DEPENDENCIES = frozenset(
    {"vb_ui", "vb_ui_makepad", "vb_ui_model", "vb_ui_snapshot", "makepad-widgets", "makepad-draw"}
)
FORBIDDEN_RUNTIME_FORMAT_DEPENDENCIES = frozenset(
    {"serde_json", "saphyr", "saphyr-parser", "serde-saphyr"}
)
DEPENDENCY_TABLES = ("dependencies", "dev-dependencies", "build-dependencies")
EXPECTED_PACKAGE_NAMES = {
    "crates/vb_boundary_inventory": "vb_boundary_inventory",
    "crates/vb_core": "vb_core",
    "crates/vb_yaml": "vb_yaml",
    "crates/vb_validate": "vb_validate",
    "crates/vb_expr": "vb_expr",
    "crates/vb_compile": "vb_compile",
    "crates/vb_storage": "vb_storage",
    "crates/vb_runtime": "vb_runtime",
    "crates/vb_doc": "vb_doc",
    "crates/vb_ipc": "vb_ipc",
    "crates/vb_codegen": "vb_codegen",
    "crates/vb_ui_makepad": "vb_ui_makepad",
    "crates/vb_ui_snapshot": "vb_ui_snapshot",
    "crates/vb_proof_kernels": "vb_proof_kernels",
    "crates/vb_cli": "velvet-ballastics",
    "crates/vb_verification": "vb_verification",
    "crates/workspace_tests": "velvet-ballastics-workspace-tests",
    "crates/vb_benchmark": "vb_benchmark",
    "fuzz": "velvet-ballastics-fuzz",
}
EXPECTED_BINARIES = {"crates/vb_cli": {"velvet-ballastics"}}
EXPECTED_FEATURES = {
    "crates/vb_core": {"default", "generated", "bench", "volatile", "test-util"},
    "crates/vb_validate": {"default", "verus"},
    "crates/vb_ui_snapshot": {"default", "std", "tokio"},
}
FORBIDDEN_FEATURE_NAMES = {"json", "serde-json", "velvet-ballistics", "velvet_ballistics"}


def load_toml(path: Path) -> dict[str, object]:
    with path.open("rb") as handle:
        loaded = tomllib.load(handle)
    return loaded


def report(message: str) -> None:
    print(message, file=sys.stderr)


def normalized_strings(value: object) -> set[str]:
    if not isinstance(value, list):
        return set()
    return {item for item in value if isinstance(item, str)}


def check_workspace_members(root: Path, failures: list[str]) -> None:
    cargo_toml = root / "Cargo.toml"
    data = load_toml(cargo_toml)
    workspace = data.get("workspace")
    if not isinstance(workspace, dict):
        failures.append("Cargo.toml: missing [workspace] table")
        return

    actual_members = normalized_strings(workspace.get("members"))
    missing_members = sorted(ACTIVE_MEMBERS - actual_members)
    extra_members = sorted(actual_members - ACTIVE_MEMBERS)
    if missing_members:
        failures.append(f"Cargo.toml: workspace.members missing {missing_members}")
    if extra_members:
        failures.append(f"Cargo.toml: workspace.members unexpected {extra_members}")

    active_deferred = sorted(actual_members & DEFERRED_MEMBERS)
    if active_deferred:
        failures.append(
            f"Cargo.toml: deferred crates must not be active workspace members {active_deferred}"
        )

    actual_excludes = normalized_strings(workspace.get("exclude"))
    missing_excludes = sorted(EXPECTED_EXCLUDES - actual_excludes)
    if missing_excludes:
        failures.append(f"Cargo.toml: workspace.exclude missing {missing_excludes}")


def package_name(manifest: dict[str, object]) -> str | None:
    package = manifest.get("package")
    if not isinstance(package, dict):
        return None
    name = package.get("name")
    if isinstance(name, str):
        return name
    return None


def binary_names(manifest: dict[str, object]) -> set[str]:
    binaries = manifest.get("bin")
    if not isinstance(binaries, list):
        return set()
    names = set()
    for binary in binaries:
        if isinstance(binary, dict):
            name = binary.get("name")
            if isinstance(name, str):
                names.add(name)
    return names


def feature_names(manifest: dict[str, object]) -> set[str]:
    features = manifest.get("features")
    if not isinstance(features, dict):
        return set()
    return {name for name in features if isinstance(name, str)}


def check_crate_names_binaries_and_features(root: Path, failures: list[str]) -> None:
    for member_path in sorted(EXPECTED_PACKAGE_NAMES):
        if member_path in DEFERRED_MEMBERS:
            continue
        manifest_path = root / member_path / "Cargo.toml"
        if not manifest_path.exists():
            failures.append(f"{member_path}/Cargo.toml: missing member manifest")
            continue
        manifest = load_toml(manifest_path)
        expected_name = EXPECTED_PACKAGE_NAMES[member_path]
        actual_name = package_name(manifest)
        if actual_name != expected_name:
            failures.append(
                f"{member_path}/Cargo.toml: package.name expected {expected_name!r}, got {actual_name!r}"
            )

        if member_path in EXPECTED_BINARIES:
            expected_binaries = EXPECTED_BINARIES[member_path]
            actual_binaries = binary_names(manifest)
            if actual_binaries != expected_binaries:
                failures.append(
                    f"{member_path}/Cargo.toml: bin names expected {sorted(expected_binaries)!r}, got {sorted(actual_binaries)!r}"
                )

        expected_features = EXPECTED_FEATURES.get(member_path)
        actual_features = feature_names(manifest)
        if expected_features is not None and actual_features != expected_features:
            failures.append(
                f"{member_path}/Cargo.toml: features expected {sorted(expected_features)!r}, got {sorted(actual_features)!r}"
            )
        forbidden_features = sorted(actual_features & FORBIDDEN_FEATURE_NAMES)
        if forbidden_features:
            failures.append(
                f"{member_path}/Cargo.toml: forbidden feature names {forbidden_features!r}"
            )


def dependency_names(manifest: dict[str, object]) -> set[str]:
    names: set[str] = set()
    for table_name in DEPENDENCY_TABLES:
        table = manifest.get(table_name)
        if isinstance(table, dict):
            for name, dependency in table.items():
                if isinstance(name, str):
                    names.add(name)
                if isinstance(dependency, dict):
                    package = dependency.get("package")
                    if isinstance(package, str):
                        names.add(package)
                    path = dependency.get("path")
                    if isinstance(path, str):
                        path_alias = path.rstrip("/\\").rsplit("/", maxsplit=1)[-1]
                        if path_alias:
                            names.add(path_alias)
    return names


def check_forbidden_dependencies(root: Path, failures: list[str]) -> None:
    for crate in sorted(BOUNDARY_CRATES):
        manifest_path = root / "crates" / crate / "Cargo.toml"
        if not manifest_path.exists():
            failures.append(f"{manifest_path.relative_to(root)}: missing boundary crate manifest")
            continue

        manifest = load_toml(manifest_path)
        names = dependency_names(manifest)

        ui_hits = sorted(names & FORBIDDEN_UI_DEPENDENCIES)
        if ui_hits:
            failures.append(
                f"{manifest_path.relative_to(root)}: forbidden UI dependency in boundary crate {crate}: {ui_hits}"
            )

        format_hits = sorted(names & FORBIDDEN_RUNTIME_FORMAT_DEPENDENCIES)
        if format_hits:
            failures.append(
                f"{manifest_path.relative_to(root)}: forbidden runtime format dependency in {crate}: {format_hits}"
            )


def check_generated_boundaries(root: Path, failures: list[str]) -> None:
    generated_roots = sorted((root / "crates").glob("*/src/generated"))
    for generated_root in generated_roots:
        for source in sorted(generated_root.rglob("*.rs")):
            text = source.read_text(encoding="utf-8")
            for forbidden in sorted(FORBIDDEN_UI_DEPENDENCIES | FORBIDDEN_RUNTIME_FORMAT_DEPENDENCIES):
                if forbidden in text:
                    failures.append(
                        f"{source.relative_to(root)}: forbidden generated boundary token {forbidden}"
                    )


def main() -> int:
    root = Path.cwd()
    failures: list[str] = []

    check_workspace_members(root, failures)
    check_crate_names_binaries_and_features(root, failures)
    check_forbidden_dependencies(root, failures)
    check_generated_boundaries(root, failures)

    for failure in failures:
        report(failure)

    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-workspace-assertions.rs
use std::collections::BTreeSet;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

const EXPECTED_MEMBERS: &[&str] = &[
    "crates/vb_boundary_inventory",
    "crates/vb_core",
    "crates/vb_yaml",
    "crates/vb_validate",
    "crates/vb_expr",
    "crates/vb_compile",
    "crates/vb_storage",
    "crates/vb_runtime",
    "crates/vb_doc",
    "crates/vb_ipc",
    "crates/vb_proof_kernels",
    "crates/vb_queue_semantics",
    "crates/vb_cli",
    "crates/vb_verification",
    "crates/vb_test_util",
    "crates/workspace_tests/idempotency_suite",
    "crates/workspace_tests",
    "crates/vb_benchmark",
];

const EXPECTED_EXCLUDES: &[&str] = &["target/miri-tmp", "crates/vb_ui", "fuzz", "crates/vb_ajc40_flux"];
const BOUNDARY_CRATES: &[&str] = &["vb_core", "vb_runtime", "vb_storage", "vb_ipc"];
const FORBIDDEN_UI_DEPENDENCIES: &[&str] = &[
    "vb_ui",
    "vb_ui_makepad",
    "vb_ui_model",
    "vb_ui_snapshot",
    "makepad-widgets",
    "makepad-draw",
];
const FORBIDDEN_RUNTIME_FORMAT_DEPENDENCIES: &[&str] =
    &["serde_json", "saphyr", "saphyr-parser", "serde-saphyr"];
const FORBIDDEN_FEATURE_NAMES: &[&str] = &[
    "json",
    "serde-json",
    "generated",
    "maxperf",
    "velvet-ballistics",
    "velvet_ballistics",
];

const EXPECTED_PACKAGE_NAMES: &[(&str, &str)] = &[
    ("crates/vb_boundary_inventory", "vb_boundary_inventory"),
    ("crates/vb_core", "vb_core"),
    ("crates/vb_yaml", "vb_yaml"),
    ("crates/vb_validate", "vb_validate"),
    ("crates/vb_expr", "vb_expr"),
    ("crates/vb_compile", "vb_compile"),
    ("crates/vb_storage", "vb_storage"),
    ("crates/vb_runtime", "vb_runtime"),
    ("crates/vb_doc", "vb_doc"),
    ("crates/vb_ipc", "vb_ipc"),
    ("crates/vb_proof_kernels", "vb_proof_kernels"),
    ("crates/vb_queue_semantics", "vb_queue_semantics"),
    ("crates/vb_cli", "velvet-ballistics"),
    ("crates/vb_verification", "vb_verification"),
    ("crates/vb_test_util", "vb_test_util"),
    (
        "crates/workspace_tests/idempotency_suite",
        "velvet-ballistics-idempotency-workspace-tests",
    ),
    (
        "crates/workspace_tests",
        "velvet-ballistics-workspace-tests",
    ),
    ("crates/vb_benchmark", "vb_benchmark"),
];

const EXPECTED_FEATURES: &[(&str, &[&str])] = &[
    (
        "crates/vb_core",
        &[
            "bench",
            "default",
            "kani-diagnostic-codes",
            "kani-vb-5iebh-check-scope",
            "kani-vb-ajc40",
            "test-util",
            "volatile",
        ],
    ),
    ("crates/vb_validate", &["default", "verus"]),
];

fn quoted_values_in_line(line: &str) -> Vec<String> {
    line.split('"')
        .enumerate()
        .filter_map(|(index, value)| (index % 2 == 1).then(|| value.to_owned()))
        .collect()
}

fn quoted_array_values(text: &str, key: &str) -> BTreeSet<String> {
    let prefix = format!("{key} = [");
    let mut active = false;
    let mut values = BTreeSet::new();

    text.lines().for_each(|line| {
        let trimmed = line.trim();
        let semantic = trimmed
            .split_once('#')
            .map_or(trimmed, |(before_comment, _comment)| before_comment.trim());
        if semantic.is_empty() {
            return;
        }
        if !active && semantic.starts_with(&prefix) {
            active = true;
        }
        if active {
            quoted_values_in_line(semantic)
                .into_iter()
                .for_each(|value| {
                    values.insert(value);
                });
            if semantic.contains(']') {
                active = false;
            }
        }
    });

    values
}

fn quoted_scalar(line: &str) -> Option<String> {
    quoted_values_in_line(line).into_iter().next()
}

fn package_name(manifest: &str) -> Option<String> {
    let mut in_package = false;
    manifest.lines().find_map(|line| {
        let trimmed = line.trim();
        if trimmed.starts_with('[') {
            in_package = trimmed == "[package]";
            return None;
        }
        (in_package && trimmed.starts_with("name ="))
            .then(|| quoted_scalar(trimmed))
            .flatten()
    })
}

fn binary_names(manifest: &str) -> BTreeSet<String> {
    let mut in_bin = false;
    let mut names = BTreeSet::new();
    manifest.lines().for_each(|line| {
        let trimmed = line.trim();
        if trimmed.starts_with('[') {
            in_bin = trimmed == "[[bin]]";
            return;
        }
        if in_bin && trimmed.starts_with("name =") {
            if let Some(name) = quoted_scalar(trimmed) {
                names.insert(name);
            }
        }
    });
    names
}

fn feature_names(manifest: &str) -> BTreeSet<String> {
    let mut in_features = false;
    let mut names = BTreeSet::new();
    manifest.lines().for_each(|line| {
        let trimmed = line.trim();
        if trimmed.starts_with('[') {
            in_features = trimmed == "[features]";
            return;
        }
        if in_features && trimmed.contains('=') {
            if let Some((name, _rest)) = trimmed.split_once('=') {
                let cleaned = name.trim();
                if !cleaned.is_empty() {
                    names.insert(cleaned.to_owned());
                }
            }
        }
    });
    names
}

fn dependency_names(manifest: &str) -> BTreeSet<String> {
    let mut in_dependencies = false;
    let mut names = BTreeSet::new();
    manifest.lines().for_each(|line| {
        let trimmed = line.trim();
        if trimmed.starts_with('[') {
            in_dependencies = matches!(
                trimmed,
                "[dependencies]" | "[dev-dependencies]" | "[build-dependencies]"
            );
            return;
        }
        if !in_dependencies || !trimmed.contains('=') || trimmed.starts_with('#') {
            return;
        }
        if let Some((name, rest)) = trimmed.split_once('=') {
            let dep_name = name.trim();
            if !dep_name.is_empty() {
                names.insert(dep_name.to_owned());
            }
            if rest.contains("package") {
                quoted_values_in_line(rest)
                    .first()
                    .cloned()
                    .into_iter()
                    .for_each(|value| {
                        names.insert(value);
                    });
            }
            if rest.contains("path") {
                quoted_values_in_line(rest)
                    .into_iter()
                    .last()
                    .into_iter()
                    .for_each(|value| {
                        if let Some(alias) = value.trim_end_matches('/').rsplit('/').next() {
                            if !alias.is_empty() {
                                names.insert(alias.to_owned());
                            }
                        }
                    });
            }
        }
    });
    names
}

fn expected_set(values: &[&str]) -> BTreeSet<String> {
    values.iter().map(|value| (*value).to_owned()).collect()
}

fn sorted_difference(left: &BTreeSet<String>, right: &BTreeSet<String>) -> Vec<String> {
    left.difference(right).cloned().collect()
}

fn push_set_failure(
    failures: &mut Vec<String>,
    label: &str,
    missing: Vec<String>,
    extra: Vec<String>,
) {
    if !missing.is_empty() {
        failures.push(format!("{label} missing {missing:?}"));
    }
    if !extra.is_empty() {
        failures.push(format!("{label} unexpected {extra:?}"));
    }
}

fn read_manifest(root: &Path, member_path: &str, failures: &mut Vec<String>) -> Option<String> {
    let manifest_path = root.join(member_path).join("Cargo.toml");
    match fs::read_to_string(&manifest_path) {
        Ok(text) => Some(text),
        Err(error) => {
            failures.push(format!(
                "{}/Cargo.toml: unreadable manifest: {error}",
                member_path
            ));
            None
        }
    }
}

fn check_workspace_members(root: &Path, failures: &mut Vec<String>) {
    let cargo_path = root.join("Cargo.toml");
    let manifest = match fs::read_to_string(&cargo_path) {
        Ok(text) => text,
        Err(error) => {
            failures.push(format!("Cargo.toml: unreadable: {error}"));
            return;
        }
    };
    let actual_members = quoted_array_values(&manifest, "members");
    let expected_members = expected_set(EXPECTED_MEMBERS);
    push_set_failure(
        failures,
        "Cargo.toml: workspace.members",
        sorted_difference(&expected_members, &actual_members),
        sorted_difference(&actual_members, &expected_members),
    );

    let actual_excludes = quoted_array_values(&manifest, "exclude");
    let expected_excludes = expected_set(EXPECTED_EXCLUDES);
    push_set_failure(
        failures,
        "Cargo.toml: workspace.exclude",
        sorted_difference(&expected_excludes, &actual_excludes),
        sorted_difference(&actual_excludes, &expected_excludes),
    );
}

fn check_crate_names(root: &Path, failures: &mut Vec<String>) {
    EXPECTED_PACKAGE_NAMES.iter().for_each(|(member_path, expected_name)| {
        if let Some(manifest) = read_manifest(root, member_path, failures) {
            let actual_name = package_name(&manifest);
            if actual_name.as_deref() != Some(*expected_name) {
                failures.push(format!(
                    "{member_path}/Cargo.toml: package.name expected {expected_name:?}, got {actual_name:?}"
                ));
            }
            if *member_path == "crates/vb_cli" {
                let actual_binaries = binary_names(&manifest);
                let expected_binaries = expected_set(&["velvet-ballistics"]);
                push_set_failure(
                    failures,
                    "crates/vb_cli/Cargo.toml: bin names",
                    sorted_difference(&expected_binaries, &actual_binaries),
                    sorted_difference(&actual_binaries, &expected_binaries),
                );
            }
            EXPECTED_FEATURES.iter().for_each(|(feature_path, expected)| {
                if member_path == feature_path {
                    let actual_features = feature_names(&manifest);
                    let expected_features = expected_set(expected);
                    push_set_failure(
                        failures,
                        &format!("{member_path}/Cargo.toml: features"),
                        sorted_difference(&expected_features, &actual_features),
                        sorted_difference(&actual_features, &expected_features),
                    );
                }
            });
            let forbidden_features = feature_names(&manifest)
                .intersection(&expected_set(FORBIDDEN_FEATURE_NAMES))
                .cloned()
                .collect::<Vec<_>>();
            if !forbidden_features.is_empty() {
                failures.push(format!(
                    "{member_path}/Cargo.toml: forbidden feature names {forbidden_features:?}"
                ));
            }
        }
    });
}

fn check_forbidden_dependencies(root: &Path, failures: &mut Vec<String>) {
    BOUNDARY_CRATES.iter().for_each(|crate_name| {
        let member_path = format!("crates/{crate_name}");
        if let Some(manifest) = read_manifest(root, &member_path, failures) {
            let names = dependency_names(&manifest);
            let ui_hits = names
                .intersection(&expected_set(FORBIDDEN_UI_DEPENDENCIES))
                .cloned()
                .collect::<Vec<_>>();
            if !ui_hits.is_empty() {
                failures.push(format!(
                    "{member_path}/Cargo.toml: forbidden UI dependency in boundary crate {crate_name}: {ui_hits:?}"
                ));
            }
            let format_hits = names
                .intersection(&expected_set(FORBIDDEN_RUNTIME_FORMAT_DEPENDENCIES))
                .cloned()
                .collect::<Vec<_>>();
            if !format_hits.is_empty() {
                failures.push(format!(
                    "{member_path}/Cargo.toml: forbidden runtime format dependency in {crate_name}: {format_hits:?}"
                ));
            }
        }
    });
}

fn collect_generated_dirs(root: &Path) -> io::Result<Vec<PathBuf>> {
    let crates_dir = root.join("crates");
    if !crates_dir.exists() {
        return Ok(Vec::new());
    }
    fs::read_dir(crates_dir)?.try_fold(Vec::new(), |mut acc, entry| {
        let path = entry?.path().join("src").join("generated");
        if path.exists() {
            acc.push(path);
        }
        Ok(acc)
    })
}

fn rust_files(root: &Path) -> io::Result<Vec<PathBuf>> {
    if !root.exists() {
        return Ok(Vec::new());
    }
    fs::read_dir(root)?.try_fold(Vec::new(), |mut acc, entry| {
        let path = entry?.path();
        if path.is_dir() {
            acc.extend(rust_files(&path)?);
        } else if path.extension().and_then(|ext| ext.to_str()) == Some("rs") {
            acc.push(path);
        }
        Ok(acc)
    })
}

fn check_generated_boundaries(root: &Path, failures: &mut Vec<String>) {
    match collect_generated_dirs(root) {
        Ok(dirs) => dirs.into_iter().for_each(|dir| match rust_files(&dir) {
            Ok(files) => files
                .into_iter()
                .for_each(|source| match fs::read_to_string(&source) {
                    Ok(text) => FORBIDDEN_UI_DEPENDENCIES
                        .iter()
                        .chain(FORBIDDEN_RUNTIME_FORMAT_DEPENDENCIES.iter())
                        .for_each(|forbidden| {
                            if text.contains(forbidden) {
                                let rel = source
                                    .strip_prefix(root)
                                    .map_or(source.as_path(), |path| path);
                                failures.push(format!(
                                    "{}: forbidden generated boundary token {forbidden}",
                                    rel.display()
                                ));
                            }
                        }),
                    Err(error) => {
                        failures.push(format!("{}: unreadable: {error}", source.display()))
                    }
                }),
            Err(error) => failures.push(format!(
                "{}: unreadable generated dir: {error}",
                dir.display()
            )),
        }),
        Err(error) => failures.push(format!("crates: unreadable: {error}")),
    }
}

fn run() -> i32 {
    let root = match std::env::current_dir() {
        Ok(path) => path,
        Err(error) => {
            eprintln!("InvalidInvocation: cannot read current directory: {error}");
            return 64;
        }
    };
    let mut failures = Vec::new();
    check_workspace_members(&root, &mut failures);
    check_crate_names(&root, &mut failures);
    check_forbidden_dependencies(&root, &mut failures);
    check_generated_boundaries(&root, &mut failures);

    failures.iter().for_each(|failure| eprintln!("{failure}"));
    if failures.is_empty() { 0 } else { 1 }
}

fn main() {
    std::process::exit(run());
}
